pkg/gce-pd-csi-driver/node.go

```diff
@@ -557,21 +566,61 @@ func (ns *GCENodeServer) NodeGetCapabilities(ctx context.Context, req *csi.NodeG
 
 func (ns *GCENodeServer) NodeGetInfo(ctx context.Context, req *csi.NodeGetInfoRequest) (*csi.NodeGetInfoResponse, error) {
 	top := &csi.Topology{
-		Segments: map[string]string{common.TopologyKeyZone: ns.MetadataService.GetZone()},
+		Segments: map[string]string{
+			common.TopologyKeyZone: ns.MetadataService.GetZone(),
+		},
 	}
 
-	nodeID := common.CreateNodeID(ns.MetadataService.GetProject(), ns.MetadataService.GetZone(), ns.MetadataService.GetName())
+	if ns.EnableDiskTopology {
+		labels, err := ns.gkeTopologyLabels(ctx, ns.MetadataService.GetName())
+		if err != nil {
+			// Perhaps we don't want to fail here.  We are introducing a new
+			// dependency and we might be better off allowing this failure to
+			// happen and moving on to retrieve the zone from GCE MDS.
+			return nil, err
+		}
 
+		for k, v := range labels {
+			top.Segments[k] = v
+		}
+	}
+
+	nodeID := common.CreateNodeID(ns.MetadataService.GetProject(), ns.MetadataService.GetZone(), ns.MetadataService.GetName())
 	volumeLimits, err := ns.GetVolumeLimits()
 
 	resp := &csi.NodeGetInfoResponse{
 		NodeId:             nodeID,
 		MaxVolumesPerNode:  volumeLimits,
 		AccessibleTopology: top,
 	}
+
+	klog.V(2).Infof("Returning NodeGetInfoResponse: %+v", resp)
+
 	return resp, err
 }
 
+// gkeTopologyLabels retrieves the node labels with the prefix
+// `topology.gke.io/` for the specified node.
+func (ns *GCENodeServer) gkeTopologyLabels(ctx context.Context, nodeName string) (map[string]string, error) {
```

Comment:
nit: this should be fetchGkeTopologyLabels or whatever to make it clear that its nontrivial.